### ISSUE TITLE
ManagedChannelBuilder.forTarget() has no compatibility check with the NR

### ISSUE THREAD

**Issue by ejona86** [grpc/grpc-java#10425]
https://github.com/grpc/grpc-java/pull/9076 and https://github.com/grpc/grpc-java/pull/9113 added unix:// support and auto-negotiation between the socket address types the NameResolver might return and the ones supported by the ManagedChannelProvider. Specifically, the `Grpc.newChannelBuilder(String, ChannelCredentials)` call was enhanced to support this, but `ManagedChannelBuilder.forTarget(String)` was not.

The mixed support was okay because the UDS NR was lower-priority than DNS, so wouldn't be used accidentally in the forTarget() API and has never worked there. But if someone is making fat jars with incorrect configuration, such that it chooses the `META-INF/services/NameResolverProvider` file from grpc-netty instead of combining the file with that from grpc-core, then the name resolver will return UDS addresses and connecting will fail with a very confusing error message:

```
Caused by: java.nio.channels.UnsupportedAddressTypeException
    at java.base/sun.nio.ch.Net.checkAddress(Unknown Source)
    at java.base/sun.nio.ch.SocketChannelImpl.connect(Unknown Source)
    at io.netty.util.internal.SocketUtils$3.run(SocketUtils.java:91)
    at io.netty.util.internal.SocketUtils$3.run(SocketUtils.java:88)
    at java.base/java.security.AccessController.doPrivileged(Native Method)
    at io.netty.util.internal.SocketUtils.connect(SocketUtils.java:88)
    at io.netty.channel.socket.nio.NioSocketChannel.doConnect(NioSocketChannel.java:322)
```

We need to give a better error message at the very least. I'm thinking we could pass the supported address types to `NameResolverRegistry.asFactory()` and have `newNameResolver()` return `null` if the provider may return types that aren't included in the supported address types. Then the error will change to "cannot find a NameResolver for foo", which is correct, because there was no DNS resolver. I could even live with hard-coding InetSocketAddress in the factory, although we should probably add `getDefaultScheme()` to the registry directly for use by `ManagedChannelRegistry`, so it can completely avoid the factory.

https://github.com/grpc/grpc-java/blob/c5c37ac51c0bd96fb8514861d45db20f383c5de2/api/src/main/java/io/grpc/NameResolverRegistry.java#L165-L172

See b/292152358 and b/292542643

**Comment by sanjaypujare** [comment]
Why does the `UdsNameResolver` even get invoked if the scheme was not `unix://` ? Does the default scheme become `unix://` when the `DnsNameResolver` is not present?

**Comment by ejona86** [comment]
> Does the default scheme become unix:// when the DnsNameResolver is not present?

Yes, the default scheme is the scheme of the highest priority NameResolver. If there is only one NameResolver (UDS), then its scheme is used.

**Comment by sanjaypujare** [comment]
> ...
> Yes, the default scheme is the scheme of the highest priority NameResolver. If there is only one NameResolver (UDS), then its scheme is used.

Interesting. Each `NameResolverProvider` implements `getDefaultScheme()` (just remembered). May be the `UdsNameResolverProvider`'s  `getDefaultScheme()` should not return it's own scheme but return `dns` instead i.e. `"unix"` should never be the default scheme - ever.

**Comment by ejona86** [comment]
Well, UDS implements `getScheme()` to return `unix`, which can't change without breaking things, and `getDefaultScheme()` calls `getScheme()`. `getDefaultScheme()` should end up getting deleted, but we haven't done so yet.

**Comment by sanjaypujare** [comment]
> and `getDefaultScheme()` calls `getScheme()`.

Actually `getScheme()` calls `getDefaultScheme()` in `NameResolverProvider` which is more confusing. The default scheme should not be the concern of a provider.

**Comment by ejona86** [comment]
Oh, okay. We were just a bit more lazy when we implemented it. In either case, `getDefaultScheme()` would be going away. But the functionality would remain similar.

**Comment by ejona86** [comment]
Fixed by #10590.
